src/functions.jl

```diff
@@ -665,6 +665,69 @@ function Base.copy(f::VectorQuadraticFunction)
     )
 end
 
+"""
+    VectorNonlinearFunction(args::Vector{Any})
+
+The vector-valued nonlinear function composed of a vector of scalar functions.
+
+## `args`
+
+The vector `args` contains the scalar elements of the nonlinear function. Each
+element must be one of the following:
+
+ * A constant value of type `T<:Real`
+ * A [`VariableIndex`](@ref)
+ * A [`ScalarAffineFunction`](@ref)
+ * A [`ScalarQuadraticFunction`](@ref)
+ * A [`ScalarNonlinearFunction`](@ref)
+
+## Example
+
+To represent the function ``f(x) = [sin(x)^2, x]``, do:
+
+```jldoctest
+julia> import MathOptInterface as MOI
+
+julia> x = MOI.VariableIndex(1)
+MOI.VariableIndex(1)
+
+julia> g = MOI.ScalarNonlinearFunction(
+           :^,
+           Any[MOI.ScalarNonlinearFunction(:sin, Any[x]), 2.0],
+       )
+^(sin(MOI.VariableIndex(1)), 2.0)
+
+julia> MOI.VectorNonlinearFunction(Any[g, x])
+┌                                 ┐
+│^(sin(MOI.VariableIndex(1)), 2.0)│
+│MOI.VariableIndex(1)             │
+└                                 ┘
+```
+"""
+struct VectorNonlinearFunction <: AbstractVectorFunction
+    rows::Vector{Any}
```

Comment:
Should we make this 
```suggestion
    rows::Vector{ScalarNonlinearFunction}
```

Reply:
@blegat, take a look at this as-is with `Any`. I was playing around with `ScalarNonlinearFunction`, and it just ends up requiring a bunch of `convert` in annoying places. If we force each row to be `+(arg)`, then the downstream solvers are still going to have to special case a bunch of stuff, and they'll still need support for `ScalarAffineFunction` etc, because they can be arguments to `ScalarNonlinearFunction`.

The only problem with the current `Any` is the subtlety around `eachscalar` and `scalarize`.

Reply:
> I was playing around with `ScalarNonlinearFunction`, and it just ends up requiring a bunch of `convert` in annoying places

Could you be more explicit ? Naively, I wouldn't think it would so I would need some examples to convince me ^^

> If we force each row to be `+(arg)`, then the downstream solvers are still going to have to special case a bunch of stuff, and they'll still need support for `ScalarAffineFunction` etc, because they can be arguments to `ScalarNonlinearFunction`.

Not really if they use `MOI.Nonlinear.Model` for AD. If they use the expression graph, they are probably still going to parse the expression and do the obvious simplifications so this simplification might just come as a special case.
It's going to cause a bunch of troubles that accessing the fields, `scalarize`, `eachscalar` and `scalar_type` don't match. I don't see an added value of that design that would outweight the added complexity of having these inconsistencies.

Reply:
Okay, take a look at this now. I think I got it to a nice point. The automatic conversion is still a bit weird. But maybe that's okay.

Reply:
[PATHSolver](https://github.com/chkwon/PATHSolver.jl/pull/82/files#diff-64cebae681efceded214317cd18b678aaefed5658af7a123f9b01433313a2844R231-R238) now needs this

<img width="634" alt="image" src="https://github.com/jump-dev/MathOptInterface.jl/assets/8177701/869c1f08-41ee-4309-af21-59f8fccc9c47">

where previously it could just require that `xi = f.rows[i+N]::MOI.VariableIndex`.

Reply:
It makes sense, it's similar and consistent with Indicator which transforms the `VariableIndex` for having it fit into the `VectorAffineFunction`. It would be nice to have tools to make it easier though.